src/inputfetcher.h

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INPUTFETCHER_H
+#define BITCOIN_INPUTFETCHER_H
+
+#include <coins.h>
+#include <sync.h>
+#include <tinyformat.h>
+#include <txdb.h>
+#include <util/threadnames.h>
+
+#include <algorithm>
+#include <iterator>
+#include <set>
+#include <vector>
+
+/**
+ * Input fetcher for fetching inputs from the CoinsDB and inserting
+ * into the CoinsTip.
+ *
+ * The main thread pushes batches of outpoints
+ * onto the queue, where they are fetched by N worker threads. The resulting
+ * coins are pushed onto another queue after they are read from disk. When
+ * the main is done adding outpoints, it starts writing the results of the read
+ * queue to the cache.
+ */
+class InputFetcher
+{
+private:
+    //! Mutex to protect the inner state
+    Mutex m_mutex{};
+
+    //! Worker threads block on this when out of work
+    std::condition_variable m_worker_cv{};
+
+    //! Main thread blocks on this when out of work
+    std::condition_variable m_main_cv{};
+
+    //! The queue of outpoints to be fetched from disk.
+    //! As the order of outpoints doesn't matter, it is used as a LIFO (stack)
+    std::vector<COutPoint> m_outpoints GUARDED_BY(m_mutex){};
+
+    //! The queue of pairs to be written to the cache.
+    std::vector<std::pair<COutPoint, Coin>> m_pairs GUARDED_BY(m_mutex){};
+
+    /**
+     * Number of outpoint fetches that haven't completed yet.
+     * This includes outpoints that are no longer queued, but still in the
+     * worker's own batches.
+     */
+    size_t m_in_flight_fetches_count GUARDED_BY(m_mutex){0};
+
+    //! The maximum number of outpoints to be processed in one batch
+    const size_t m_batch_size;
+
+    //! DB to fetch from.
+    const CCoinsViewDB* m_db{nullptr};
+
+    std::vector<std::thread> m_worker_threads;
+    bool m_request_stop GUARDED_BY(m_mutex){false};
+
+    /** Internal function that does the fetching from disk. */
+    void Loop() noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        size_t in_flight_fetches_count{0};
+        std::vector<std::pair<COutPoint, Coin>> pairs{};
+        do {
+            std::vector<COutPoint> outpoints{};
+            outpoints.reserve(m_batch_size);
+            {
+                WAIT_LOCK(m_mutex, lock);
+                // first do the clean-up of the previous loop run (allowing us to do it in the same critsect)
```

Comment:
What kind of sect is a critsect?
```suggestion
                // first do the clean-up of the previous loop run (allowing us to do it in the same critical section)
```